src/hypothesis/strategies.py

```diff
@@ -836,31 +836,78 @@ def dm_func(denom):
 
 @cacheable
 @defines_strategy
-def decimals(
-    min_value=None, max_value=None, allow_nan=None, allow_infinity=None
-):
-    """Generates instances of decimals.Decimal.
+def decimals(min_value=None, max_value=None,
+             allow_nan=True, allow_infinity=None, places=None):
+    """Generates instances of decimals.Decimal, which may be:
 
-    - If min_value is not None, all values will be >= min_value.
-    - If max_value is not None, all values will be <= max_value.
-    - If min_value or max_value is not None, it is an error to enable
-      allow_nan.
-    - If both min_value and max_value are not None, it is an error to enable
-      allow_infinity.
+    - A finite rational number, between ``min_value`` and ``max_value``.
+    - Not a Number, if ``allow_nan`` is truthy.
+    - Positive or negative infinity, if ``max_value`` and ``min_value``
+      respectively are None, and ``allow_infinity`` is not False.  (None means
+      "allow infinity, unless excluded by the min and max values")
 
-    Where not explicitly ruled out by the bounds, all of infinity, -infinity
-    and NaN are possible values generated by this strategy.
+    Note that where floats have one `NaN` value, Decimals have four: signed,
+    and either *quiet* or *signalling*.  See `the decimal module docs
+    <https://docs.python.org/3/library/decimal.html#special-values>`_ for
+    more information on special values.
+
+    If ``places`` is not None, all finite values drawn from the strategy will
+    have that number of digits after the decimal place.
 
     """
-    floats_strategy = floats(
-        min_value=min_value, max_value=max_value,
-        allow_nan=allow_nan, allow_infinity=allow_infinity)
-    fractions_strategy = fractions(min_value=min_value, max_value=max_value)
-
-    return (
-        floats_strategy.map(float_to_decimal) |
-        fractions_strategy.map(lambda f: Decimal(f.numerator) / f.denominator)
-    )
+    # Convert min_value and max_value to Decimal values, and validate args
+    check_valid_integer(places)
+    if places is not None and places < 0:
+        raise InvalidArgument('cannot have negative number of places (%r)')
+    if min_value is not None:
+        min_value = Decimal(min_value)
+        if math.isinf(min_value) and min_value < 0:
+            min_value = None
+    if max_value is not None:
+        max_value = Decimal(max_value)
+        if math.isinf(max_value) and max_value > 0:
+            max_value = None
+    if min_value and not min_value.is_finite():
```

Comment:
We actually still want to exclude `min_value='infinity'`, because we won't be able to draw any finite values (and the infinity with an allowed sign was converted to None above).  And `check_valid_bound` uses `math.isnan`, which is lovely but doesn't work for `Decimal('sNaN')`, which doesn't have a floating-point equivalent.